packages/frame-core/src/types.ts

```diff
@@ -26,14 +27,117 @@ export type FrameContext = {
      * Profile image URL
      */
     pfpUrl?: string;
-  }
+  };
 };
 
+export const notificationDetailsSchema = z.object({
+  url: z.string(),
+  token: z.string(),
+});
+export type FrameNotificationDetails = z.infer<
+  typeof notificationDetailsSchema
+>;
+
+export type AddFrameResult =
+  | {
+      added: true;
+      notificationDetails?: FrameNotificationDetails;
+    }
+  | {
+      added: false;
+      reason: "invalid-domain-manifest" | "rejected-by-user";
+    };
+
+export type AddFrame = () => Promise<AddFrameResult>;
+
 export type FrameHost = {
   context: FrameContext;
   close: () => void;
   ready: () => void;
   openUrl: (url: string) => void;
   setPrimaryButton: SetPrimaryButton;
   ethProviderRequest: EthProviderRequest;
+  addFrame: AddFrame;
 };
+
+// Webhook event format (= JSON Farcaster Signature)
+
+export const eventSchema = z.object({
+  header: z.string(),
+  payload: z.string(),
+  signature: z.string(),
+});
+export type EventSchema = z.infer<typeof eventSchema>;
+
+// Webhook event header after decoding
+
+export const eventHeaderSchema = z.object({
+  fid: z.number(),
+  type: z.literal("app_key"),
+  key: z.string(),
+});
+export type EventHeader = z.infer<typeof eventHeaderSchema>;
+
+// Webhook event payload after decoding
+
+export const eventFrameAddedPayloadSchema = z.object({
+  event: z.literal("frame-added"),
+  notificationDetails: notificationDetailsSchema.optional(),
+});
+export type EventFrameAddedPayload = z.infer<
+  typeof eventFrameAddedPayloadSchema
+>;
+
+export const eventFrameRemovedPayloadSchema = z.object({
+  event: z.literal("frame-removed"),
+});
+export type EventFrameRemovedPayload = z.infer<
+  typeof eventFrameRemovedPayloadSchema
+>;
+
+export const eventNotificationsEnabledPayloadSchema = z.object({
+  event: z.literal("notifications-enabled"),
+  notificationDetails: notificationDetailsSchema.required(),
+});
+export type EventNotificationsEnabledPayload = z.infer<
+  typeof eventNotificationsEnabledPayloadSchema
+>;
+
+export const notificationsDisabledPayloadSchema = z.object({
+  event: z.literal("notifications-disabled"),
+});
+export type EventNotificationsDisabledPayload = z.infer<
+  typeof notificationsDisabledPayloadSchema
+>;
+
+export const eventPayloadSchema = z.discriminatedUnion("event", [
+  eventFrameAddedPayloadSchema,
+  eventFrameRemovedPayloadSchema,
+  eventNotificationsEnabledPayloadSchema,
+  notificationsDisabledPayloadSchema,
+]);
+export type FrameEvent = z.infer<typeof eventPayloadSchema>;
+
+// Notifications API request and response formats
+
+export const sendNotificationRequestSchema = z.object({
+  notificationId: z.string(),
+  title: z.string(),
+  body: z.string(),
+  targetUrl: z.string(),
+  tokens: z.array(z.string()),
```

Comment:
should enforce a max here?

Reply:
good point, added a limit of 100, also added limits to a few other places